Add Base32Kit

jenslauterbach opened the pull request:
This PR will add `Base32` encoding and decoding to this repository. It is all the code from my repository without any restructuring that might be warranted.

Lets discuss restructuring later :)

jenslauterbach committed:
Add Base32Kit

jenslauterbach committed:
Switch to actions/checkout@v2

jenslauterbach committed:
Change Docker Volume Mount

The current mount does not work. The `$GITHUB_WORKSPACE` environment
variable is not interpolated in the `ci.yml` (at least not for the
`volume` option of the container).

According to the "internet" the workspace is mounted anyway to
"/home/workspace" in the container itself.

So I changed the "workdir" of the container to "$GITHUB_WORKSPACE" and
hope this will fix the issue.

jenslauterbach committed:
Remove Container Options

jenslauterbach committed:
Change Codecov Action Version

jenslauterbach committed:
Add 'curl' to build container

This is necessary because the 'ubuntu' container used in the action does
not contain 'curl'. Without 'curl' the codecov action is not able to
upload the result.

fabianfett commented on .github/workflows/ci.yaml:
```diff
@@ -18,20 +18,19 @@ jobs:
-      options: --workdir /src
     steps:
     - name: Checkout
-      uses: actions/checkout@v1
```
I like these changes, but I would prefer them to be in a separate PR: Normally I prefer small prs with one single purpose. @jenslauterbach are you willing to open another pr for that? We could also include the nightly 5.3 image already.

fabianfett commented on Package.swift:
```diff
@@ -9,6 +9,9 @@ let package = Package(
     .library(
       name: "Base64Kit",
       targets: ["Base64Kit"]),
+    .library(
+      name: "Base32Kit",
+      targets: ["Base32Kit"])
```
As I stated before: I'm unsure if we want to create a separate target for Base32. Other option would be the renaming of our current target: `Base64Kit` to `BaseKit`. @jenslauterbach do you have arguments for one or the other on that?

jenslauterbach replied:
I thought about `BaseKit` but I suspect that it might be confusing. One could think that it is something like a `Commons` package in other language like Java. Just stuff that is _commonly_ needed or something. Even though it is a bit more verbose, we might want something like `BaseEncodingKit`?

jenslauterbach replied:
> do you have arguments for one or the other on that?

No. My gut tells me it would be better to have a single target. Naming will be the challenge 😄

fabianfett replied:
@jenslauterbach Okay. Then let's just merge everything into Base64Kit first, and we can do a big renaming pr later.

jenslauterbach replied:
@fabianfett Great. I am will make the required changes.

fabianfett commented on Sources/Base32Kit/Alphabets.swift:
```diff
@@ -0,0 +1,120 @@
+struct Alphabet {
+    static let encodePaddingCharacter: UInt8 = UInt8(ascii: "=")
```
You use four spaces indention, whereas my code uses just two spaces. Since my coding guidelines are just in my head, I will create a separate pr where we will start to use `swiftformat`, hence no discussion.

jenslauterbach replied:
Alright :)

fabianfett commented on Sources/Base32Kit/Alphabets.swift:
```diff
@@ -0,0 +1,120 @@
+
+extension Base32 {
+
+    private static let __: UInt8 = 255
```
👍

fabianfett commented on Sources/Base32Kit/Base32Kit.swift:
```diff
@@ -0,0 +1,375 @@
+/// 1. The Base 32 Alphabet (see [RFC 4648 Section 6](https://tools.ietf.org/html/rfc4648#section-6))
+/// 2. The "Extended Hex" Base 32 Alphabet (see [RFC 4648 Section 7](https://tools.ietf.org/html/rfc4648#section-7))
+///
+/// The API only supports the encoding and decoding of strings and no other data types. All the inputs are validated
```
> The API only supports the encoding and decoding of strings and no other data types.

Why is that? We limit ourselves to quiet a subset here?

jenslauterbach replied:
I only needed `String`. But I agree with your argument that any array of `Byte` should be supported. For now I would make this an improvement and change that in another PR, if that's ok for you. I still have a few things I would like to implement/change. But I wanted to keep this PR to a minimum.

fabianfett replied:
I don't think that we should add this in a later pr. Anything we merge into master should be as good as possible.

jenslauterbach replied:
👍 Understood. Working on it.

fabianfett commented on Sources/Base32Kit/Base32Kit.swift:
```diff
@@ -0,0 +1,375 @@
+    ///
+    /// - Returns: Base 32 Hex encoded `String` or empty `String` if the given `string` is empty.
+    ///
+    /// - Important: This method is case insensitive.
```
> Important: This method is case insensitive.

How is that meant?

fabianfett commented on Sources/Base32Kit/Base32Kit.swift:
```diff
@@ -0,0 +1,375 @@
+    public static func encodeHex(string: String) -> String {
+        if string.isEmpty {
+            return ""
+        }
```
Is this necessary?

fabianfett commented on Sources/Base32Kit/Base32Kit.swift:
````diff
@@ -0,0 +1,375 @@
+/// ```
+///
+/// - Note: Encoding and decoding methods are not optimized and might perform badly. Use another Swift package if
+///         performance is a primary concern.
````
We need to fix that. We don't write slow code...

fabianfett commented on Sources/Base32Kit/Validation.swift:
```diff
@@ -0,0 +1,130 @@
+    ///        if encoded string contains one or more illegal characters.
+    ///     - `Base32.DecodingError.invalidPaddingCharacters`
+    ///        if encoded string contains a padding character (`=`) at an illegal position.
+    static func validate(string: String, legalCharacters: String) throws {
```
this should be part of the decoding effort.

fabianfett commented on Sources/Base32Kit/Base32Kit.swift:
```diff
@@ -0,0 +1,375 @@
+        static let strictMode: DecodingOptions = [.requiredPadding]
+    }
+
+    private static func decode<Buffer: Collection>(bytes: Buffer, alphabet: [UInt8]) throws -> String
```
The generic here doesn't make any sense if we are only operating on `[UInt8]` anyways. But we should make this public.

fabianfett commented on Sources/Base32Kit/Base32Kit.swift:
```diff
@@ -0,0 +1,375 @@
+            output.append(value)
+        }
+
+        return String(decoding: output, as: Unicode.UTF8.self)
```
Decoding into a String doesn't make sense IMHO. The bytes can be anything between 0...255. Forcing those into a `String` limits the versatilely of the method.

jenslauterbach replied:
Same as before. I only needed String. Ok if I change that in a later PR? Overall, we should make sure that all the base encoding methods that are offered have the same API "structure". Would not make sense if the base 32 methods take string and the base 64 methods take bytes.

fabianfett commented on Sources/Base32Kit/Errors.swift:
```diff
@@ -0,0 +1,40 @@
+extension Base32 {
+
+    /// Errors that can be thrown during decoding of a Base 32 encoded string.
+    public enum DecodingError: Error, Equatable {
```
Those error types should probably be merged with their base64 counterparts.

jenslauterbach replied:
Definitely 👍

fabianfett replied:
If you rebase off master you'll find it already in a separate file.

fabianfett commented on Sources/Base32Kit/Validation.swift:
```diff
@@ -0,0 +1,130 @@
+    /// - Returns: true if the given `string` contains invalid padding, otherwise false.
+    ///
+    /// - Complexity: O(n)
+    static func invalidPadding(in string: String) -> Bool {
```
Why do we need this? I assume this should be part of the decoding.

jenslauterbach replied:
It is only required in the decoding. But validation became quite big so I moved it to a separate file/`struct`. Not yet accustomed to Swift "style". `SwiftLint` was complaining about file length etc.

fabianfett replied:
I have never used `SwiftLint` before. I meant do we need this at all? The check whether something is valid Base32 is implicitly in the decoding (at least this is how Base64 works).

jenslauterbach replied:
> I meant do we need this at all

To be honest: I am not 100% sure. In general my decoding made sure that the given "string" is valid before decoding was done. In theory you probably can check for invalid padding during encoding. But the way it is implemented at the moment I am not sure that all invalid paddings can be covered without look ahead/behind in the byte array or checking where exactly in the array the decoding is at the moment.

But I saw no implementation that did it like this, so I would guess that my implementation is wrong. I will check again.

fabianfett commented on Tests/Base32KitTests/XCTestManifests.swift:
```diff
@@ -0,0 +1,11 @@
+import XCTest
```
We don't need this, since we are only supporting 5.1 and above

jenslauterbach replied:
I did not know that. Will change it.

fabianfett commented on Tests/LinuxMain.swift:
```diff
@@ -1 +1,7 @@
+import XCTest
 
+import Base32KitTests
+
+var tests = [XCTestCaseEntry]()
+tests += Base32KitTests.allTests()
+XCTMain(tests)
```
Since we are supporting 5.1 and above only this file can remain empty.

jenslauterbach replied:
Ok.

fabianfett reviewed (changes requested):
Thanks so much for the contribution! 👍 I'm sure we can evolve this into something really great. There are a couple of things though that we need to address. 

1. Change ci setup in different pr. (#14)
2. Code style. I will enable `swiftformat` on this repo and all code should be in this style. (#14)
3. Performance. This base64 implementation was initially created to be used on the server (and offer a better alternative to the Data methods while not linking Foundation). Performance therefore is a must. We will need to test this against the base64 implementation (where it should be slower. But we need to determine the margin).
4. Count mallocs. This is a task for @fabianfett. We should enable test for base32 as well. (#16)
5. Merge functionality and utils. Right now we have the same error twice `DecodingError`. This is probably a task for myself. Restructure the current repo to make room for different Base implementations. (#15)
6. String only. This does not make any sense to me at all... Of course we encode into a `String` and decode from a `String` but the encoding input can be any number of bytes in the range from 0 to 255. The same applies for the output. The default encoding input should therefore be `[UInt8]` and the default decoding output should also be `[UInt8]`.

jenslauterbach committed:
Merge remote-tracking branch 'upstream/master' into add-base32

jenslauterbach committed:
Apply 'swiftformat' formatting

jenslauterbach committed:
Clean LinuxMain.swift

jenslauterbach committed:
Remove 'XCTestManifests.swift'

jenslauterbach committed:
Remove Base32Kit Target

In agreement with Fabian I refactored the code so that there is only one
target left. For now we keep the name "Base64Kit". Changing the name
will be part of a later pull request.

To be able to move my code into the Base64Kit folders I had to do some
renaming.

jenslauterbach committed:
Remove Unnecessary 'allTests' Variables

jenslauterbach committed:
Rename Base32Kit to be in line with Base64Kit

jenslauterbach committed:
Merge branch 'master' into add-base32

jenslauterbach committed:
Merge branch 'master' into add-base32

jenslauterbach committed:
Major Refactoring of API

As discussed with Fabian, we refactor the API to allow the consumer to
select the desired alphabet. For now we only support "standard" and
"hex". In the future, I'd like to add even more alphabets like
crockford and bech32.

jenslauterbach committed:
Remove 'Errors.swift'

Fabian and I decided to use a single, common 'DecodingError' struct for
all decoding errors.

jenslauterbach committed:
Remove 'Errors.swift'

jenslauterbach committed:
Remove commented code

jenslauterbach committed:
Make 'Base32' struct public

gwynne commented:
This functionality was implemented in #29
